_posts/2016-03-10-information-hiding.md

```diff
@@ -0,0 +1,334 @@
+
+It is important that such abstractions must be done carefully.
+Often times a type is visible to the caller, and if the type
+is not handled carefully the abstraction can leak.
+
+For example, a set with fast random indexing (useful for random
+walks on a graph) can be implemented with a sorted `Vector`.
+However, if the `Vector` type is
+leaked, the user can use this knowledge to violate the invariant.
+
+```scala
+import scala.annotation.tailrec
+
+def binarySearch(i: Int, repr: Vector[Int]): (Boolean, Int) = {
+  @tailrec
+  def binarySearchAux(minIndex: Int, maxIndex: Int): (Boolean, Int) = {
+    val midIndex = minIndex + ((maxIndex - minIndex) / 2)
+
+    if (maxIndex < minIndex) (false, midIndex)
+    else {
+      if (repr(midIndex) > i) binarySearchAux(minIndex, midIndex - 1)
+      else if (repr(midIndex) < i) binarySearchAux(midIndex + 1, maxIndex)
+      else (true, midIndex)
+    }
+  }
+  binarySearchAux(0, repr.size - 1)
+}
+
+object IntSet {
+  type Repr = Vector[Int]
+
+  def empty: Repr = Vector.empty
+
+  def add(i: Int, repr: Repr): Repr = {
+    val (isMember, indexOf) = binarySearch(i, repr)
+    if (isMember) repr
+    else {
+      val (prefix, suffix) = repr.splitAt(indexOf)
+      prefix ++ Vector(i) ++ suffix
+    }
+  }
+
+  def contains(i: Int, repr: Repr): Boolean =
+    binarySearch(i, repr)._1
+}
+```
+
+```scala
+import IntSet._
+// import IntSet._
+
+val good = add(1, add(10, add(5, empty)))
+// good: IntSet.Repr = Vector(1, 5, 10)
+
+val goodResult = contains(10, good)
+// goodResult: Boolean = true
+
+val bad = good.reverse // We know it's a Vector!
+// bad: scala.collection.immutable.Vector[Int] = Vector(10, 5, 1)
+
+val badResult = contains(10, bad)
+// badResult: Boolean = false
+
+val bad2 = Vector(10, 5, 1) // Alternatively..
+// bad2: scala.collection.immutable.Vector[Int] = Vector(10, 5, 1)
+
+val badResult2 = contains(10, bad2)
+// badResult2: Boolean = false
+```
+
+The issue here is the user knows more about the representation than they
+should. The function `add` enforces the sorted invariant on each insert,
+and the function `contains` leverages this to do an efficient look-up.
+Because the `Vector` definition of `Repr` is exposed, the user is
+free to create any `Vector` they wish which may violate the invariant,
+thus breaking `contains`.
+
+In general, the **name** of the representation type is needed but the
+**definition** is not. If the definition is hidden, the user is only able to
+work with the type to the extent the module allows. This is precisely
+the notion of information hiding. If this can be enforced by the type
+system, modules can be swapped in and out without worrying about breaking
+client code.
+
+# Quantification
+It turns out there is a [well understood principle][understandingTypes]
+behind this idea called *existential quantification*. Contrast with
+universal quantification which says "for all", existential quantification
+says "there exists."
```

Comment:
I usually prefer to say "there is a" instead of "there exists", because otherwise you'd explain "existential" by "exists". (That's just my 2¢ as a non-native speaker.)